### PR TITLE
Version 3.1

### PR DESCRIPTION
**Pull request by zyware**
* Updated to 3.0.0-ALPHA10
* Fixed all known bugs.

### CODE REVIEW

**Comment by zyware** [comment]
Not yet ready, do not close.

**Comment by zyware** [comment]
`[17:00:12] [Server thread/INFO]: zzyware joined the game
[17:00:44] [Server thread/CRITICAL]: Unhandled exception executing command 'cs' in chestshop: Undefined property: pocketmine\nbt\tag\CompoundTag::$ChestShop
[17:00:44] [Server thread/CRITICAL]: ErrorException: "Undefined property: pocketmine\nbt\tag\CompoundTag::$ChestShop" (EXCEPTION) in "plugins/ChestShop-master/src/ChestShop/Main" at line 225
[17:00:44] [Server thread/DEBUG]: #0 plugins/ChestShop-master/src/ChestShop/Main(225): pocketmine\error_handler(integer 8, string Undefined property: pocketmine\nbt\tag\CompoundTag::$ChestShop, string C:\Users\mikha\Desktop\PocketMine-MP\plugins\ChestShop-master\src\ChestShop\Main.php, integer 225, array Array())
[17:00:44] [Server thread/DEBUG]: #1 plugins/ChestShop-master/src/ChestShop/Main(173): ChestShop\Main->fillInventoryWithShop(ChestShop\Chest\CustomChestInventory object)
[17:00:44] [Server thread/DEBUG]: #2 plugins/ChestShop-master/src/ChestShop/Main(401): ChestShop\Main->sendChestShop(pocketmine\Player object)
[17:00:44] [Server thread/DEBUG]: #3 src/pocketmine/command/PluginCommand(58): ChestShop\Main->onCommand(pocketmine\Player object, pocketmine\command\PluginCommand object, string cs, array Array())
[17:00:44] [Server thread/DEBUG]: #4 src/pocketmine/command/SimpleCommandMap(258): pocketmine\command\PluginCommand->execute(pocketmine\Player object, string cs, array Array())
[17:00:44] [Server thread/DEBUG]: #5 src/pocketmine/Server(1989): pocketmine\command\SimpleCommandMap->dispatch(pocketmine\Player object, string cs)
[17:00:44] [Server thread/DEBUG]: #6 src/pocketmine/Player(2117): pocketmine\Server->dispatchCommand(pocketmine\Player object, string cs)
[17:00:44] [Server thread/DEBUG]: #7 src/pocketmine/network/mcpe/PlayerNetworkSessionAdapter(220): pocketmine\Player->chat(string /cs)
[17:00:44] [Server thread/DEBUG]: #8 src/pocketmine/network/mcpe/protocol/CommandRequestPacket(54): pocketmine\network\mcpe\PlayerNetworkSessionAdapter->handleCommandRequest(pocketmine\network\mcpe\protocol\CommandRequestPacket object)
[17:00:44] [Server thread/DEBUG]: #9 src/pocketmine/network/mcpe/PlayerNetworkSessionAdapter(92): pocketmine\network\mcpe\protocol\CommandRequestPacket->handle(pocketmine\network\mcpe\PlayerNetworkSessionAdapter object)
[17:00:44] [Server thread/DEBUG]: #10 src/pocketmine/network/mcpe/protocol/BatchPacket(118): pocketmine\network\mcpe\PlayerNetworkSessionAdapter->handleDataPacket(pocketmine\network\mcpe\protocol\CommandRequestPacket object)
[17:00:44] [Server thread/DEBUG]: #11 src/pocketmine/network/mcpe/PlayerNetworkSessionAdapter(92): pocketmine\network\mcpe\protocol\BatchPacket->handle(pocketmine\network\mcpe\PlayerNetworkSessionAdapter object)
[17:00:44] [Server thread/DEBUG]: #12 src/pocketmine/Player(2964): pocketmine\network\mcpe\PlayerNetworkSessionAdapter->handleDataPacket(pocketmine\network\mcpe\protocol\BatchPacket object)
[17:00:44] [Server thread/DEBUG]: #13 src/pocketmine/network/mcpe/RakLibInterface(149): pocketmine\Player->handleDataPacket(pocketmine\network\mcpe\protocol\BatchPacket object)
[17:00:44] [Server thread/DEBUG]: #14 vendor/pocketmine/raklib/server/ServerHandler(103): pocketmine\network\mcpe\RakLibInterface->handleEncapsulated(string 192.168.0.87 22033, raklib\protocol\EncapsulatedPacket object, integer 0)
[17:00:44] [Server thread/DEBUG]: #15 src/pocketmine/network/mcpe/RakLibInterface(88): raklib\server\ServerHandler->handlePacket()
[17:00:44] [Server thread/DEBUG]: #16 src/pocketmine/network/Network(89): pocketmine\network\mcpe\RakLibInterface->process()
[17:00:44] [Server thread/DEBUG]: #17 src/pocketmine/Server(2534): pocketmine\network\Network->processInterfaces()
[17:00:44] [Server thread/DEBUG]: #18 src/pocketmine/Server(2285): pocketmine\Server->tick()
[17:00:44] [Server thread/DEBUG]: #19 src/pocketmine/Server(2158): pocketmine\Server->tickProcessor()
[17:00:44] [Server thread/DEBUG]: #20 src/pocketmine/Server(1744): pocketmine\Server->start()
[17:00:44] [Server thread/DEBUG]: #21 src/pocketmine/PocketMine(385): pocketmine\Server->__construct(BaseClassLoader object, pocketmine\utils\MainLogger object, string C:\Users\mikha\Desktop\PocketMine-MP\, string C:\Users\mikha\Desktop\PocketMine-MP\plugins\)`

**Comment by zyware** [comment]
I better completely rewrite the plugin 😄

File: src/ChestShop/EventListener.php [hunk 1]
```diff
@@ -73,6 +74,10 @@ public function onTransaction(InventoryTransactionEvent $event) : void{
 		$playerAction = null;
 
 		foreach($actions as $action){
+			if($action instanceof DropItemAction) {
+				$event->setCancelled(true);
```
Comment (Muqsit) [inline src/ChestShop/EventListener.php]: Why are you cancelling the event completely?

File: src/ChestShop/Main.php [hunk 2]
```diff
@@ -106,7 +106,7 @@ public function onEnable() : void{
 			$this->updateConfig($file);
 		}
 
-		$shops = yaml_parse_file($this->getDataFolder().'shops.yml');
+		$shops = @yaml_parse_file($this->getDataFolder().'shops.yml');
```
Comment (Muqsit) [inline src/ChestShop/Main.php]: What's the reason for this?

**Comment by Muqsit** [comment]
The NBT error you're facing is due to the recent changes on how CompoundTag works. I'm closing this. I'm thinking of using InvMenu for handling ChestShop transactions.

File: src/ChestShop/EventListener.php (hunk 1 above)
Comment (zyware) [reply src/ChestShop/EventListener.php]: @Muqsit Why not?

File: src/ChestShop/Main.php (hunk 2 above)
Comment (zyware) [reply src/ChestShop/Main.php]: @Muqsit https://hastebin.com/qafoketipo.tex

Comment (Muqsit) [reply src/ChestShop/Main.php]: https://github.com/Muqsit/ChestShop/blob/master/src/ChestShop/Main.php#L102

File: src/ChestShop/EventListener.php (hunk 1 above)
Comment (Muqsit) [reply src/ChestShop/EventListener.php]: You are cancelling all `DropItemAction` in the process, even when the player isn't viewing the custom inventory.

**Comment by zyware** [comment]
@Muqsit How soon will you do this? Or do not tell me how I can do this?

**Pull request closed without merging**